{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finding a Protein Motif \n",
    "Given: At most 15 UniProt Protein Database access IDs.\n",
    "Return: For each protein possessing the N-glycosylation motif, output its given access ID followed by a list of locations in the protein string where the motif can be found."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# for this excercise i got this error\n",
    "# urllib.error.URLError: <urlopen error [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:997)>\n",
    "# and i was not able to fix it \n",
    "# that's why on Rosalind i didn't complete the problem\n",
    "# but i think it work if i could fix this error\n",
    "\n",
    "\n",
    "import urllib.request as url\n",
    "import re\n",
    "from urllib.request import urlopen\n",
    "from Bio import SeqIO \n",
    "\n",
    "ID = [] \n",
    "\n",
    "proteins = 'P00740_FA9_HUMAN\\nP07987_GUX2_TRIRE\\nP01042_KNH_HUMAN\\nP39873_RNBR_BOVIN\\nP11171_41_HUMAN\\nQ3BN23\\nA6UUD2\\nP20840_SAG1_YEAST\\nQ3ATP6\\nB5FPF2\\nP01047_KNL2_BOVIN\\nA6LJ74\\nB5FNU0'\n",
    "\n",
    "protein_list = proteins.split('\\n')\n",
    "\n",
    "for protein in protein_list:\n",
    "    \n",
    "    link = \"http://www.uniprot.org/uniprot/\"+ protein +\".FASTA\"\n",
    "    data = urlopen(link)                                       \n",
    "    fasta = data.read().decode('utf-8', 'ignore')             \n",
    "    with open('seq_file.fasta', 'a') as text_file:            \n",
    "        text_file.write(fasta)                                \n",
    "                                                              \n",
    "    handle = open('seq_file.fasta', 'r')                          \n",
    "    motifs = re.compile(r'(?=(N[^P][ST][^P]))')                   \n",
    "    count = 0                                                     \n",
    "    for record in SeqIO.parse(handle, 'fasta'):                   \n",
    "        sequence = record.seq                                     \n",
    "        positions = []                                            \n",
    "    for m in re.finditer(motifs, str(sequence)):              \n",
    "                positions.append(m.start() + 1)                               \n",
    "    if len(positions) > 0:                                    \n",
    "        print(ID[count])                                      \n",
    "        print(' '.join(map(str, positions)))                  \n",
    "        count += 1          "
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Open Reading Frames\n",
    "Either strand of a DNA double helix can serve as the coding strand for RNA transcription. Hence, a given DNA string implies six total reading frames, or ways in which the same region of DNA can be translated into amino acids: three reading frames result from reading the string itself, whereas three more result from reading its reverse complement.\n",
    "\n",
    "An open reading frame (ORF) is one which starts from the start codon and ends by stop codon, without any other stop codons in between. Thus, a candidate protein string is derived by translating an open reading frame into amino acids until a stop codon is reached.\n",
    "\n",
    "Given: A DNA string s of length at most 1 kbp in FASTA format.\n",
    "\n",
    "Return: Every distinct candidate protein string that can be translated from ORFs of s. Strings can be returned in any order."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# I don't know how many attempts i made a this problem but i managed to do it right in the end\n",
    "\n",
    " \n",
    "def translate(rnaseq):\n",
    "    codon_table = { 'UUU': 'F', 'CUU': 'L', 'AUU': 'I', 'GUU': 'V', \\\n",
    "                    'UUC': 'F', 'CUC': 'L', 'AUC': 'I', 'GUC': 'V', \\\n",
    "                    'UUA': 'L', 'CUA': 'L', 'AUA': 'I', 'GUA': 'V', \\\n",
    "                    'UUG': 'L', 'CUG': 'L', 'AUG': 'M', 'GUG': 'V', \\\n",
    "                    'UCU': 'S', 'CCU': 'P', 'ACU': 'T', 'GCU': 'A', \\\n",
    "                    'UCC': 'S', 'CCC': 'P', 'ACC': 'T', 'GCC': 'A', \\\n",
    "                    'UCA': 'S', 'CCA': 'P', 'ACA': 'T', 'GCA': 'A', \\\n",
    "                    'UCG': 'S', 'CCG': 'P', 'ACG': 'T', 'GCG': 'A', \\\n",
    "                    'UAU': 'Y', 'CAU': 'H', 'AAU': 'N', 'GAU': 'D', \\\n",
    "                    'UAC': 'Y', 'CAC': 'H', 'AAC': 'N', 'GAC': 'D', \\\n",
    "                    'UAA': 'Stop', 'CAA': 'Q', 'AAA': 'K', 'GAA': 'E', \\\n",
    "                    'UAG': 'Stop', 'CAG': 'Q', 'AAG': 'K', 'GAG': 'E', \\\n",
    "                    'UGU': 'C', 'CGU': 'R', 'AGU': 'S', 'GGU': 'G', \\\n",
    "                    'UGC': 'C', 'CGC': 'R', 'AGC': 'S', 'GGC': 'G', \\\n",
    "                    'UGA': 'Stop', 'CGA': 'R', 'AGA': 'R', 'GGA': 'G', \\\n",
    "                    'UGG': 'W', 'CGG': 'R', 'AGG': 'R', 'GGG': 'G'}\n",
    "\n",
    "    length = len(rnaseq)\n",
    "    proseq = []\n",
    "    for i in range(0,length,3):\n",
    "        triplet = rnaseq[i:i+3]\n",
    "        if codon_table[str(triplet)] != 'Stop':\n",
    "            proseq.append(codon_table[str(triplet)])\n",
    "        else:\n",
    "            break\n",
    "    proseq = ''.join(proseq)\n",
    "    return proseq\n",
    "\n",
    "def rev_comp(x):\n",
    "    tran = ''\n",
    "    for i in range(len(x)):\n",
    "        index = -(i+1)\n",
    "        if x[index] == 'A':\n",
    "            tran = tran + 'U'\n",
    "        elif x[index] == 'T':\n",
    "            tran = tran + 'A'\n",
    "        elif x[index] == 'C':\n",
    "            tran = tran + 'G'\n",
    "        else:\n",
    "            tran = tran +'C'\n",
    "    revcomp = ''\n",
    "    for i in range(len(tran)):\n",
    "        index = -(i+1)\n",
    "        if tran[index] == 'A':\n",
    "            revcomp = revcomp + 'U'\n",
    "        elif tran[index] == 'U':\n",
    "            revcomp = revcomp + 'A'\n",
    "        elif tran[index] == 'C':\n",
    "            revcomp = revcomp + 'G'\n",
    "        else:\n",
    "            revcomp = revcomp +'C'\n",
    "    return tran, revcomp\n",
    " \n",
    "def find_all(s,substring):\n",
    "    index_list = []\n",
    "    index = s.find(substring)\n",
    "    while index != -1:\n",
    "        index_list.append(index)\n",
    "        index = s.find(substring, index+1)\n",
    "    if len(index_list) > 0:\n",
    "        return index_list\n",
    "    else:\n",
    "        return -1\n",
    " \n",
    "def orf(mrna):\n",
    "    start_codon = 'AUG'\n",
    "    stop_codon = ['UAA', 'UAG', 'UGA']\n",
    "    i, j = 0,0\n",
    "    out = []\n",
    "    while i <= len(mrna)-2:\n",
    "        if mrna[i:i+3] == start_codon:\n",
    "            j=i\n",
    "            sequence=''\n",
    "            while i<= len(mrna) -2:\n",
    "                if mrna[i:i+3] in stop_codon:\n",
    "                    out.append(sequence)\n",
    "                    break\n",
    "                sequence = sequence + mrna[i:i+3]\n",
    "                i = i+3\n",
    "        i = j+1\n",
    "        j = j+1\n",
    "    return out  \n",
    "\n",
    "dna = ''\n",
    "mrna1, mrna2 = rev_comp(dna)\n",
    "orf_list1=orf(mrna1)\n",
    "orf_list2=orf(mrna2)\n",
    "orf_list = orf_list1\n",
    "for i in orf_list2:\n",
    "    if i not in orf_list:\n",
    "        orf_list.append(i)\n",
    "for i in orf_list:\n",
    "    print(translate(i))\n",
    " "
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Overlapping Graphs\n",
    "A graph whose nodes have all been labeled can be represented by an adjacency list, in which each row of the list contains the two node labels corresponding to a unique edge.\n",
    "\n",
    "A directed graph (or digraph) is a graph containing directed edges, each of which has an orientation. That is, a directed edge is represented by an arrow instead of a line segment; the starting and ending nodes of an edge form its tail and head, respectively. The directed edge with tail v and head w is represented by (v,w) (but not by (w,v)). A directed loop is a directed edge of the form (v,v).\n",
    "\n",
    "For a collection of strings and a positive integer k, the overlap graph for the strings is a directed graph Ok in which each string is represented by a node, and string s is connected to string t with a directed edge when there is a length k suffix of s that matches a length k prefix of t, as long as s≠t; we demand s≠t to prevent directed loops in the overlap graph (although directed cycles may be present).\n",
    "\n",
    "Given: A collection of DNA strings in FASTA format having total length at most 10 kbp.\n",
    "\n",
    "Return: The adjacency list corresponding to O3. You may return edges in any order."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "file = 'rosalind_grph.txt'\n",
    "\n",
    "with open(file) as f:\n",
    "    sequences = list(SeqIO.parse(f, 'fasta'))\n",
    "\n",
    "for fasta1 in sequences:\n",
    "    for fasta2 in sequences:\n",
    "        name1, sequence1 = fasta1.id, str(fasta1.seq)\n",
    "        name2, sequence2 = fasta2.id, str(fasta2.seq)\n",
    "        if sequence1 == sequence2:\n",
    "            continue\n",
    "        suffix = sequence1[-3:]\n",
    "        prefix = sequence2[:3]\n",
    "        if prefix == suffix:\n",
    "            print(name1, name2)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "RNA Splicing\n",
    "After identifying the exons and introns of an RNA string, we only need to delete the introns and concatenate the exons to form a new string ready for translation.\n",
    "\n",
    "Given: A DNA string s (of length at most 1 kbp) and a collection of substrings of s acting as introns. All strings are given in FASTA format.\n",
    "\n",
    "Return: A protein string resulting from transcribing and translating the exons of s. (Note: Only one solution will exist for the dataset provided.)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rna2aa_dictionary =  {'UUU' : 'F', \"UUC\": 'F', 'UUA' : 'L', 'UUG' : 'L', 'UCU' : 'S' , 'UCA' :'S', 'UCC' : 'S', 'UCG' : 'S', 'UAU' : 'Y', 'UAC': 'Y', 'UAA': 'STOP' , 'UAG': 'STOP', 'UGU' : 'C', 'UGC': 'C', 'UGA' : 'STOP', 'UGG': 'W', 'CUU' : 'L', 'CUC' : 'L', 'CUA': 'L', 'CUG': 'L', 'CCU': 'P', 'CCC' : 'P', 'CCA' : 'P', 'CCG' : 'P', 'CAU' : 'H', 'CAC':'H', 'CAA':'Q','CAG':'Q','CGU': 'R','CGC':'R','CGA':'R','CGG':'R','AUU':'I','AUC':'I','AUA':'I','AUG':'M','ACU':'T','ACC':'T','ACA':'T','ACG':'T','AAU':'N','AAC':'N','AAA':'K','AAG':'K','AGU':'S','AGC':'S','AGA':'R','AGG':'R','GUU':'V','GUC':'V','GUA':'V','GUG':'V','GCU':'A','GCC':'A','GCA':'A','GCG':'A','GAU':'D','GAC':'D','GAA':'E','GAG':'E','GGU':'G','GGC':'G','GGA':'G','GGG':'G'}\n",
    "\n",
    "file = open(\"rosalind_splc.txt\", \"r\")\n",
    "\n",
    "def extract_sequence(file):\n",
    "    sequences = []\n",
    "    result = \"\"       \n",
    "    \n",
    "    for seq in file:\n",
    "        if \">\" in seq:\n",
    "            sequences.append(result)\n",
    "            result = \"\"\n",
    "        else:\n",
    "            if \"\\n\" in seq:\n",
    "                result += seq[:len(seq) - 1]\n",
    "            else:\n",
    "                result += seq\n",
    "\n",
    "    sequences.append(result)\n",
    "    sequences.remove('')\n",
    "    return sequences\n",
    "\n",
    "sequences = extract_sequence(file)\n",
    "\n",
    "dna_string = sequences[0]\n",
    "introns = sequences[1:]\n",
    "introns.sort(reverse = True)\n",
    "\n",
    "for intron in introns:\n",
    "    dna_string = dna_string.replace(intron,\"\")\n",
    "dna_string = dna_string.replace(\"T\",\"U\")\n",
    "\n",
    "for i in range(0, len(dna_string) - 3, 3):\n",
    "    print(rna2aa_dictionary[dna_string[i: i +3]],end='')      \n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finding a spliced Motif\n",
    "A subsequence of a string is a collection of symbols contained in order (though not necessarily contiguously) in the string (e.g., ACG is a subsequence of TATGCTAAGATC). The indices of a subsequence are the positions in the string at which the symbols of the subsequence appear; thus, the indices of ACG in TATGCTAAGATC can be represented by (2, 5, 9).\n",
    "\n",
    "As a substring can have multiple locations, a subsequence can have multiple collections of indices, and the same index can be reused in more than one appearance of the subsequence; for example, ACG is a subsequence of AACCGGTT in 8 different ways.\n",
    "\n",
    "Given: Two DNA strings s and t (each of length at most 1 kbp) in FASTA format.\n",
    "\n",
    "Return: One collection of indices of s in which the symbols of t appear as a subsequence of s. If multiple solutions exist, you may return any one."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "fasta_formatted_input = \">Rosalind_9160\\nGATCGATAGGCAATTTGGCTCGGGAGAATTCTTGGTGTGCTCACTCCGATTCTCTTCCGATTAAGAGTTTGGTTCCTTGGGCAATGCAGCTTGCTCATGGGGGAAGATTAAGGGGAACTTTTGGCTATGAGACATAGACAGACCTTAGTGTGGCGTCTAGTCGGAATCTAGAGCCACACTAGTGTACATAATCGGGTGAGACATTACCGAAGACGGATCGCTGTTCGGGTAATGATGAATCCTGAAGATTGTCGTCTATGCTATGCTTCCGTTGCTCGGATCCGGGACTGACAAAAAAGCAGGGGAGGTGTTTAAAGGCTAAGCCTCCAAAAGGTGCAGAACATGACGAGCCTTCTATCGGGTAAGTGTTAAGGCGTCGTCCAGAATCTCGCATCTCGCTCTACGACTCTCGTTTGAAAACCAGTTGACGGACGCTCCTAGCCGCTACCGGAGGACATAATCTGGTTCAGGCGCCACCTACAGGTTCTGACACGTGGGACTTTAACTCGCCCCTAAAATCAGTAGACGAAACCGGTAAGGGTTCATTTCGGCCAAGGCGATGCACATCATCTACGGTTCCGAGGCTGGTGCTCGTCACGGCGATCCTAGATACCGCGACACCAGTGCAACCTGTGTGGTAGTGAAGAGGGCTGAGTCGACGCGCGGCGATCGCGTCGAAAACTACTTCCAGTACAACCAGGAAACCCTGGTGAAATTGTTTTGATAAACCCTTATATTTTCATGGTTAGGGCAAGAATGAACCCTAAGCCTTACACCCATCTGTGTCAACGGCGTTAGTGGCACTTAGTACCTGCCGTAACTGTGATATAAAGGTTTGTGAAGGAAAGGCCCACACTCTCTCGTTCTGGAGCGTTGACCAGAATGGCCAATCGGCTGATGCGCGCGCTACATGACATACGCACAAGTCTTTGCGATCGTATG\\n>Rosalind_1858\\nTTGCGGATGCTCCGCTAGGGTACCTCTCCCTTCCGAAGGTGCGGCTCACTCTAGCTCGTTT\"\n",
    "splitted_input = fasta_formatted_input.split(\"\\n\")\n",
    "\n",
    "dna_sequence = splitted_input[1]\n",
    "motif = splitted_input[3]\n",
    "\n",
    "indices = \"\"\n",
    "dna_index = 0\n",
    "motif_index = 0\n",
    "\n",
    "for base in dna_sequence:\n",
    "    if base == motif[motif_index]:\n",
    "        motif_index += 1\n",
    "        indices += str(dna_index + 1) + \" \"\n",
    "        if motif_index == len(motif):\n",
    "            break\n",
    "        \n",
    "    dna_index += 1\n",
    "print(indices)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Transitions and Transversions\n",
    "For DNA strings s1 and s2 having the same length, their transition/transversion ratio R(s1,s2) is the ratio of the total number of transitions to the total number of transversions, where symbol substitutions are inferred from mismatched corresponding symbols as when calculating Hamming distance (see “Counting Point Mutations”).\n",
    "\n",
    "Given: Two DNA strings s1 and s2 of equal length (at most 1 kbp).\n",
    "\n",
    "Return: The transition/transversion ratio R(s1,s2)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2.0106382978723403\n"
     ]
    }
   ],
   "source": [
    "fasta = \">Rosalind_8091\\nTCCACGCCGCTTACGACTCATTGGTGATTTTGACTAGACGAGTAGTGAAAAAGCTATTCTCCGTCAACAGATGAGATGTTAGATAGATTGTTGTTCCTTTTCACTGTCAATTGGTATTGACTCCCCTTCCAAACCTTTTAGCATACATGACTGTCCTAGCTCCAGTACCGGTAGTGTAGAGTCAGACCCTCAAACATAAGTAAGAGACGATTTATTGACCGTAGCCGAGTCCAGCATAGTTTTTATCATTACGATTTCGATAAGATTATGCAAGGCCTCGCCTCGGGTAGAACGTATAGATGCTCTAACCGAGTTCCATGTCGGGACACACGCCGATCACACTAAACAGTATCATTTTATCTTAAAGTGAGTCCTGGCTGCTACGGGATTATGCGCGCGGGTCACACGATCGAGAGACAGTACAAGCACACGAAAGAGACACATGCGCAAGCCCACCGCGGAACGACAAATATTTGTAATACGCCATAGCGGCTTATCCAACCGGAATGCGCGAAGCATGCCACGCTGCTCCAAGGCTCGAGTTCCGACACTCAACTCAACGGACGGGGGGGCTGCTAGGCTCTAACCCGGCTGACGCAAGGATGGCACCCGGAGAACCACGCATTTGTTCCTGGTCTTAACTAACTTAGGCATTGCTTTCAGTTACGGCCAGCATCTTTTCAGAGTCTCCCTGGGTGTAGCATACCCAAAACAACTGGTGTTCGCCTCAAGGTAGCGTAGGAATAACTTCAGAGAAATTACTACCCAGCGACTCGTACCGGGAGTGTTAGGTTCGAAATTGCCCGCACCGCTGTGGGCTAGGGGCACATATATGGCTTTTCTTTGCATACCGGAGTTCACGGTTAACGGCTACCCAAGCTCTTAATGCTTGAATCGTAACGCCAACCCTGCCCGAAGCTGGGATTGTTGTATGGTAATTGAA\\n>Rosalind_3665\\nTCCACACCGCTAACAACTCACTAGTCACTCCGACTCGACAAGTGGTGTAGTAGATAACCTCCGTCAACAGAGGAGATGTTGGATAAATCGTACTTCCTTATTACTGTCAGGTGGTATTGATACCCCTTCGGAACCTTATAGTAGACGTGGGCATTCTAGTTCCAGCGCTAACACTATCGAGCTGCACTCTCAATCATGAGTAAGAGACAGCTTACTAACCGTGGCTGAGTCCGGGTTATTTTGAATGTTTACGATTTCGATGAGATTATGTGAAGCTTGACTTCGAAGAAAACGCCTAGATGCTCTAACCGGGTTCCATATCCCGAGTCATGCTGATCACATTAACTACTACCAATCGATCTTGGAACAAGTCCGGGCCCCTGCACAATGATGCACACGAGTCATACGATCGGGAGACAGTACAAACACGTGAAAAGGCAACGTATGCAAGTCCACCGCAAAGTAACATAAATTGGTAATGCGCCATAGCGGCTTACTTAACCAGAATCTGCGAGGCACACCGGGTTTTTCCAAGGGCCGAGCCCCGAAAGTCGATTCGAGGGACGGGGGGGCTGCTAGGGTCAAATCCGGTTGACGTAGAGGGGGTGACGGACGAACCACGCAGCCGTGCCAGACCTTGACTAACTTAGGTATTATTGTAAGCCACGGCCATTATCTTTACAGGGTTTCTCTGGGTGTGATGTACCCAAAATTACCGGCGTACGCCTCAAAACAGTTCAGGTATAGACTCACAGAAATTGCTCCCAAGCGACATGCACCGGGAGCGCCAGGTGCGGAACTACACGTGCCGCAATCAGCTCGGAGTACATGTATCATTTACTTTCGCAAGCCGGAGTTCACGGTTAACAGCTATTCAGAGTCTCAGTGATTGCATCCTGACGCTATCCTTGCTTGAAACTGGGAAGGTCGCACCGCAGCTAAC\"\n",
    "splitted = fasta.split(\"\\n\")\n",
    "first_string = splitted[1]\n",
    "second_string = splitted[3]\n",
    "\n",
    "def tran(base1, base2):\n",
    "    return (base1 + base2) in [\"AG\",\"GA\",\"CT\",\"TC\"]\n",
    "\n",
    "transition = 0\n",
    "transversion = 0\n",
    "\n",
    "for i in range(len(first_string)):\n",
    "    first_base, second_base = first_string[i], second_string[i]\n",
    "    \n",
    "    if first_base != second_base:\n",
    "        if tran(first_base, second_base):\n",
    "            transition += 1\n",
    "        else:\n",
    "            transversion += 1\n",
    "\n",
    "ratio = (transition / transversion)\n",
    "\n",
    "print(ratio)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Completing a Tree\n",
    "An undirected graph is connected if there is a path connecting any two nodes. A tree is a connected (undirected) graph containing no cycles; this definition forces the tree to have a branching structure organized around a central core of nodes, just like its living counterpart. See Figure 2.\n",
    "\n",
    "We have already grown familiar with trees in “Mendel's First Law”, where we introduced the probability tree diagram to visualize the outcomes of a random variable.\n",
    "\n",
    "In the creation of a phylogeny, taxa are encoded by the tree's leaves, or nodes having degree 1. A node of a tree having degree larger than 1 is called an internal node.\n",
    "\n",
    "Given: A positive integer n (n≤1000) and an adjacency list corresponding to a graph on n nodes that contains no cycles.\n",
    "\n",
    "Return: The minimum number of edges that can be added to the graph to produce a tree."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "33\n"
     ]
    }
   ],
   "source": [
    "# I had to try any times because i couldn't figure out why it wasn't working \n",
    "# but then I noticed that the problem was the -1 missing\n",
    "\n",
    "input = \"rosalind_tree.txt\"\n",
    "\n",
    "with(open(input, 'r')) as f:\n",
    "    n = int(f.readline())\n",
    "    nodes = [line.split() for line in f]\n",
    "\n",
    "edges = (n - len(nodes) - 1)\n",
    "print(edges)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Genome Assembly as Shortest Superstring\n",
    "For a collection of strings, a larger string containing every one of the smaller strings as a substring is called a superstring.\n",
    "\n",
    "By the assumption of parsimony, a shortest possible superstring over a collection of reads serves as a candidate chromosome.\n",
    "\n",
    "Given: At most 50 DNA strings of approximately equal length, not exceeding 1 kbp, in FASTA format (which represent reads deriving from the same strand of a single linear chromosome).\n",
    "\n",
    "The dataset is guaranteed to satisfy the following condition: there exists a unique way to reconstruct the entire chromosome from these reads by gluing together pairs of reads that overlap by more than half their length.\n",
    "\n",
    "Return: A shortest superstring containing all the given strings (thus corresponding to a reconstructed chromosome)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# I found this code pretty difficult so I helped myself with the internet \n",
    "\n",
    "from Bio import SeqIO\n",
    "import io\n",
    "\n",
    "def get_superstring(reads_list, superstring=''):\n",
    "    if len(reads_list) == 0:\n",
    "        return superstring\n",
    "\n",
    "    elif len(superstring) == 0:\n",
    "        superstring = reads_list.pop(0)\n",
    "        return get_superstring(reads_list, superstring)\n",
    "\n",
    "    else:\n",
    "        for current_read_index in range(len(reads_list)):\n",
    "            current_read = reads_list[current_read_index]\n",
    "            current_read_length = len(current_read)\n",
    "\n",
    "            for trial in range(current_read_length // 2):\n",
    "                overlap_length = current_read_length - trial\n",
    "\n",
    "                if superstring.startswith(current_read[trial:]):\n",
    "                    reads_list.pop(current_read_index)\n",
    "                    return get_superstring(reads_list, current_read[:trial] + superstring)\n",
    "\n",
    "                if superstring.endswith(current_read[:overlap_length]):\n",
    "                    reads_list.pop(current_read_index)\n",
    "                    return get_superstring(reads_list, superstring + current_read[overlap_length:])\n",
    "\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "\n",
    "    small_dataset = \"\"\"\n",
    "\"\"\"\n",
    "    handle = io.StringIO(small_dataset)\n",
    "\n",
    "    reads = []\n",
    "    for record in SeqIO.parse(handle, 'fasta'):\n",
    "        reads.append(str(record.seq))\n",
    "    handle.close()\n",
    "    ov = get_superstring(reads)\n",
    "    print(ov)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Partial Permutations\n",
    "A partial permutation is an ordering of only k objects taken from a collection containing n objects (i.e., k≤n). For example, one partial permutation of three of the first eight positive integers is given by (5,7,2).\n",
    "\n",
    "The statistic P(n,k) counts the total number of partial permutations of k objects that can be formed from a collection of n objects. Note that P(n,n) is just the number of permutations of n objects, which we found to be equal to n!=n(n−1)(n−2)⋯(3)(2) in “Enumerating Gene Orders”.\n",
    "\n",
    "Given: Positive integers n and k such that 100≥n>0 and 10≥k>0.\n",
    "\n",
    "Return: The total number of partial permutations P(n,k), modulo 1,000,000."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "n = 93\n",
    "k = 10\n",
    "\n",
    "partial_perm = 1\n",
    "for i in range(k):\n",
    "    partial_perm *= (n - i)\n",
    "print(partial_perm % 1000000)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Perfect Matchings and RNA Secondary Structures\n",
    "\n",
    "Given: An RNA string s of length at most 80 bp having the same number of occurrences of 'A' as 'U' and the same number of occurrences of 'C' as 'G'.\n",
    "\n",
    "Return: The total possible number of perfect matchings of basepair edges in the bonding graph of s."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO                      \n",
    "from math import factorial \n",
    "\n",
    "sequence = ''                              \n",
    "fasta = open('rosalind_pmch.txt', 'r')     \n",
    "for x in SeqIO.parse(fasta, 'fasta'):\n",
    "    sequence = str(x.seq)             \n",
    "fasta.close()                             \n",
    "\n",
    "AU = 0                                     \n",
    "GC = 0                                     \n",
    "for k in sequence:                        \n",
    "    if k == 'A':                          \n",
    "        AU += 1                            \n",
    "    elif k == 'G':                        \n",
    "        GC += 1                            \n",
    "\n",
    "match = factorial(AU) * factorial(GC)  \n",
    "print(match)               "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.0"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "aee8b7b246df8f9039afb4144a1f6fd8d2ca17a180786b69acc140d282b71a49"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
